common/usermode.c

```diff
@@ -124,13 +124,43 @@ static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, lo
     return return_code;
 }
 
+inline long __user_text sysenter(long syscall_nr, long arg1, long arg2, long arg3,
+                                        long arg4, long arg5) {
+    register long return_code asm(STR(_ASM_AX));
+    register long _arg4 asm("r8") = arg4;
+    register long _arg5 asm("r9") = arg5;
+    register long _arg3 asm("r10") = arg3;
+
+    /* clang-format off */
+    asm volatile (
+        SAVE_CALLEE_SAVED_REGS()
+        "mov %%" STR(_ASM_SP) ", %%" STR(_ASM_CX) "\n"
+        "mov $1f, %%" STR(_ASM_DX) "\n"
+        "sysenter\n"
+        "1: " RESTORE_CALLEE_SAVED_REGS()
+        : "=a"(return_code)
+        : "0"(syscall_nr), "S"(arg1), "D" (arg2), "r" (_arg3), "r"(_arg4), "r"(_arg5)
+        : STR(_ASM_CX), STR(_ASM_DX)
+    );
+    /* clang-format on */
+
+    return return_code;
+}
+
 #define syscall0(nr)                     syscall((nr), 0, 0, 0, 0, 0)
 #define syscall1(nr, a1)                 syscall((nr), (a1), 0, 0, 0, 0)
 #define syscall2(nr, a1, a2)             syscall((nr), (a1), (a2), 0, 0, 0)
 #define syscall3(nr, a1, a2, a3)         syscall((nr), (a1), (a2), (a3), 0, 0)
 #define syscall4(nr, a1, a2, a3, a4)     syscall((nr), (a1), (a2), (a3), (a4), 0)
 #define syscall5(nr, a1, a2, a3, a4, a5) syscall((nr), (a1), (a2), (a3), (a4), (a5))
 
+#define sysenter0(nr)                     sysenter((nr), 0, 0, 0, 0, 0)
+#define sysenter1(nr, a1)                 sysenter((nr), (a1), 0, 0, 0, 0)
+#define sysenter2(nr, a1, a2)             sysenter((nr), (a1), (a2), 0, 0, 0)
+#define sysenter3(nr, a1, a2, a3)         sysenter((nr), (a1), (a2), (a3), 0, 0)
+#define sysenter4(nr, a1, a2, a3, a4)     sysenter((nr), (a1), (a2), (a3), (a4), 0)
+#define sysenter5(nr, a1, a2, a3, a4, a5) sysenter((nr), (a1), (a2), (a3), (a4), (a5))
+
```

Comment:
Hmm, do we need those? Using `SYSCALL` or `SYSENTER` (or `INT $vec` if we support this someday) is an "implementation detail" of the syscall ABI and not meant to be used directly -- especially as the various instructions have all their specialties to handle.

I think an generic `syscall(...)` API is sufficient with the implementation details all handled within this function. However, to support testing different instructions, a "syscall mode switch" should be provided, e.g. something like this:

```c
enum syscall_mode {
  SYSCALL_MODE_SYSCALL,     // use SYSCALL
  SYSCALL_MODE_SYSENTER,    // use SYSENTER
  SYSCALL_MODE_INT,         // TODO: use INT $SYSVEC
};

bool syscall_mode(enum syscall_mode);
```

Reply:
This is good idea!